### PR TITLE
findById Optimization

### PR DESCRIPTION
**Pull request by TheovanKraay**
# Description
The current default behaviour of `findById(ID id)` is that it will always execute a cross partition key query. This is not an optimal behaviour for containers with big datasets/large numbers of partitions. This change is an optimization to make `findById(ID id)` execute a point read when partitionKey is `id`, and log a warning when it is a field other than id. The new default behaviour is almost certain to be what user intends. Current behaviour can potentially cause significant unintended latency issues.

# All SDK Contribution checklist:
- [x] **The pull request does not introduce [breaking changes]**
- [x] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md).**

## [General Guidelines and Best Practices](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#developer-guide)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-java/blob/main/CONTRIBUTING.md#building-and-unit-testing)
- [x] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/core/CosmosTemplateIT.java
Comment (trande4884) [inline sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/core/CosmosTemplateIT.java:222]: This doesn't seem to be testing anything unique to show that it is in fact doing the point read and not cross-partition, is there anything to test to be sure it isn't doing a cross-partition query? Maybe here we can test that the warning message added isn't logged and in testFindById() below we can test that the warning message is logged? The same goes for the reactive tests.

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (trande4884) [inline sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:357]: I believe this will have some issues. You can make any field the Id by adding @Id above it so there is no guarantee that the name of the id field will be "id". You will want to create a CosmosEntityInformation objet and use the getIdFieldName() function instead of just putting in "id". Message me if you need me to show you how to do this :)

File: sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/domain/BasicItem.java
Comment (trande4884) [inline sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/domain/BasicItem.java:1]: Add the following at the top of all new files, which can be found in any file to copy/paste:

// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

File: sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/domain/BasicItem.java
Comment (trande4884) [inline sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/domain/BasicItem.java:14]: Add another type for testing where the name of this field is something other than "id" to ensure everything is working properly.

Comment (TheovanKraay) [reply sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/domain/BasicItem.java:14]: The @Id notation here is actually redundant/misleading, and I should have left it out. The “id” field will be defined as Id by default. It doesn’t matter what we define as id, only whether the id field is defined as @partitionKey

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (TheovanKraay) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:357]: Here we’re looking for the field that has been defined as the partition key for the container. It’s true you can define any field as “Id” but that doesn’t dictate what has been defined as the partitionKey from my understanding.

Comment (TheovanKraay) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:357]: So if I define “lastName” as id, as far as I understand the value of “lastName” will be used to populate “id”. But if I’ve said “id” is the @partitionKey, then the result of this assignment will still be “id”, and that’s what’s being checked here. What’s crucial is whether Id has been defined as partitionKey. One thing I don’t know is whether id is defined as partitionKey by default if nothing is defined for partition key. Will have to test.

File: sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/core/CosmosTemplateIT.java
Comment (TheovanKraay) [reply sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/core/CosmosTemplateIT.java:222]: Well, this test is using a new entity where id is explicitly defined as partitionKey. Since that’s the only scenario in which findById will execute a point read, it will have to be a point read. But yeah, adding a test for the warning message might be better.

**Comment by TheovanKraay** [comment]
> @TheovanKraay will this change be ported to the spring boot 3 branch?

Oh I had not intended to as I expected this to be merged earlier as it’s only a small change. Can you pull from main when merged? Or will Spring Boot 3 branch be merged very soon?

File: sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/domain/BasicItem.java
Comment (TheovanKraay) [reply sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/domain/BasicItem.java:1]: Ah, is that why builds are failing on this file? Ok, will do, thanks!

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (trande4884) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:357]: I think it is worth testing, if you have:

@Id
@PartitionKey
String lastName;

It may just be putting lastName in a field called "id" but I don't think that is the case, I think "lastName" is the only field created but even if it is put in field "id" I still think when you get the name of partition key it will be "lastName".  A lot of room for things to regress regardless with so many different pieces so worth adding the test where the field is not named "id" if you ask me.

Comment (trande4884) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java]: Discussed offline with Theo, no action needed.

**Review by trande4884 — APPROVED** [review]

**Review by trande4884 — APPROVED** [review]

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/ReactiveCosmosTemplate.java
Comment (trande4884) [inline sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/ReactiveCosmosTemplate.java:337]: I think this is causing some of your integration issues. We should call this above containerProperties and pass finalContainerName into getContainerProperties().

Comment (TheovanKraay) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/ReactiveCosmosTemplate.java:334]: Ditching getContainerProperties(), see above

**Comment by saragluna** [comment]
> > @TheovanKraay will this change be ported to the spring boot 3 branch?
> 
> Oh I had not intended to as I expected this to be merged earlier as it’s only a small change. Can you pull from main when merged? The changes are very small. Or will Spring Boot 3 branch be merged very soon?

The Spring Boot 3 branch has been set up, there will be a weekly or bi-weekly sync from the main branch. However, the code under sdk/spring folder won't be synced in that process, because our code has been diverged from the main branch. So we have to create another PR to cherry-pick the changes to the sb3 branch.

**Comment by TheovanKraay** [comment]
> > > @TheovanKraay will this change be ported to the spring boot 3 branch?
> > 
> > 
> > Oh I had not intended to as I expected this to be merged earlier as it’s only a small change. Can you pull from main when merged? The changes are very small. Or will Spring Boot 3 branch be merged very soon?
> 
> @TheovanKraay, the Spring Boot 3 branch has been set up, there will be a weekly or bi-weekly sync from the main branch. However, the code under sdk/spring folder won't be synced in that process, because our code has been diverged from the main branch. So we have to create another PR to cherry-pick the changes to the sb3 branch.

Ok, we will either create another PR to cherry pick the changes, or ditch this and just create the PR to sb3 branch.

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (FabianMeiswinkel) [inline sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:365]: Logging a warning is a breaking change - and doesn't really help - like discussed before if we want to raise visibility the right way would eb to deprecate the findById method altogether (to raise warning at compile time vs. runtime)

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (FabianMeiswinkel) [inline sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:357]: What if id PK value in that case. is null - please map to the correct special

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (trande4884) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:365]: findById() is a Spring method which we cannot deprecate, even though this is a breaking change I believe it is a good approach moving forward.

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (trande4884) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:359]: I don't believe we need to wrap it in getStringIDValue, we can just pass id into the new PartitionKey()

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (trande4884) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:356]: cosmosEntityInformation is tied to the Spring domain class, where PK is tagged with @PartitionKey so all attributes like field name are pulled from the local domain class. This is how we grab such information in Spring.

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (TheovanKraay) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:359]: Agreed, will change, thanks.

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (TheovanKraay) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:357]: There could only be `new PartitionKey(null)` if id was also null, since this path is only called when id is also the pk... so `findById(id, domainType, new PartitionKey(id))` would fail in this case anyway.

**Comment by trande4884** [comment]
@TheovanKraay please update the change log also

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (kushagraThapar) [inline sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:361]: This is a breaking change, as currently customers get `CosmosAccessException` and they must have a handling for it, in case if something goes bad. Now we are logging and returning null and not throwing the exception. We should not change this behavior.

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/ReactiveCosmosTemplate.java
Comment (kushagraThapar) [inline sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/ReactiveCosmosTemplate.java:336]: This cannot return null. The API expects a Mono<T> - a reactive type. We cannot return null here. 
Also, catching the exception is incorrect here.

**Review by kushagraThapar — CHANGES_REQUESTED** [review]
Thanks @TheovanKraay  for the changes, while the changes look good in general, we should also add tests for cases where customers directly use the repository API - `findById(ID id)` and make sure we get the point read item done. 

Also, on the current tests, we should make sure there are no feed response diagnostics (to make sure it is not a query rather a point read). Thanks!

File: sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/repository/repository/BasicItemRepository.java
Comment (kushagraThapar) [inline sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/repository/repository/BasicItemRepository.java:1]: missing license header

File: sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/repository/repository/ReactiveBasicItemRepository.java
Comment (kushagraThapar) [inline sdk/spring/azure-spring-data-cosmos/src/test/java/com/azure/spring/data/cosmos/repository/repository/ReactiveBasicItemRepository.java:1]: missing license header

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (TheovanKraay) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:361]: Ok, will reset on this.

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (FabianMeiswinkel) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:365]: We have already seen CRIs when elevating log noise level. In many production environments logging with INFO is enabled - warnings/errors even more frequently. Logging a new message for every single method invocation is adding overhead (and making logs much harder to reason about). So, it is a perf and behavioral regression. FindById could easily be executed hundreds or thousands of times per second. What is the value in logging this warning over-and-over again. Why not log a warning or error only once?

Comment (FabianMeiswinkel) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:365]: @TheovanKraay - I am absolutely fine with the behavior change (it is an improvement - no regression I am seeing there) - except for the logging. Customers having a working app right now where their log files aren't swamped with warnings might get hundreds or thousands of pretty much identical log lines after just upgrading the sdk. What value doe we provide my logging this for every method invocation?

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (FabianMeiswinkel) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:356]: Interesting - makes sense we only need to know how to extarct the partition key value - not how to build the hash - so, Hash V1 vs. V2 is irrelevant.

Side question - do we have @PartitionKey support for hierarchical partition keys yet?

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (FabianMeiswinkel) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:365]: I think warning is fine as long as you only log it once per process - basically keep some state whether this warning was logged or not - that  way you don't cause too much noise, but still have the warning level logs. Do you consider that acceptable?

Comment (TheovanKraay) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:365]: Sounds fine to me. Kushagra wants some other changes too, will work on that tomorrow.

Comment (TheovanKraay) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:365]: Actually I don’t know if it makes sense to log only once per process. My thinking is that, if an app uses findById frequently, logging only once might be missed and not even noticed, so we may as well not have logged it at all. In one sense the whole idea is that if they are doing a cross pk query the “noise” of it being logged again and again causes them to actually change their app. The noise in their logs is better than the effect of doing cross pk queries if they have large containers and never realise until raising a support case. If they have small containers and/or low throughput, then the logging noise won’t matter so much anyway (but it still might cause them to change their app to remove the noise, and they’ll still be making the app more efficient in the end). So actually I think we should leave as WARN. My assumption is that many customers will not have WARN level in prod environments, but they will have it in pre-prod, so they’ll see this immediately and repeatedly when they upgrade versions, and that’s what we want. But I’m unsure what’s typical. @kushagraThapar @trande4884 any thoughts on this?

Comment (FabianMeiswinkel) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:365]: Logs are emitted at runtime - when it is already too late to warn a developer. When making changes that require customers to change their application to consume a new SDK version we loose their trust in any new update. And in this case we know for sure that many customers use the findById for months - even the CRI I talked about yesterday (initially 10 partitions for few hundred documents) ran that service happily for several months. Starting to log >1000 log lines per second will force them to make changes to the application. And IMO that isn't justified. Either break  it at compile time and document it as a breaking change or avoid breaking the customer (as in don't force them to change their app). Single log per process I can live with because it nudges the cx and maybe some pick-up on it to improve their experience - but at least it doesn't break.

Comment (FabianMeiswinkel) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:365]: We also have evidence showing that cx are not happy with increasing noisy   warning logs - Annie had added a log recently when channels got closed under certain non-happy-path circumstances - which also led to this warning being emitted for timeouts - we had 2 CRIs for it. We had similar CRIs for certain warning in RNTBD stack on certain error conditions.

**Comment by TheovanKraay** [comment]
> Thanks @TheovanKraay for the changes, while the changes look good in general, we should also add tests for cases where customers directly use the repository API - `findById(ID id)` and make sure we get the point read item done.
> 
> Also, on the current tests, we should make sure there are no feed response diagnostics (to make sure it is not a query rather a point read). Thanks!

I found there was ample coverage for point read (i.e. executing `findByID (ID id)` where id is also partition key) in the integration tests, and also coverage of cross partition key read with `findById (ID id)` in reactive tests, and these all seem to pass. I did add one test for the cross partition case in `AddressRepositoryIT`

**Comment by kushagraThapar** [comment]
> 

@TheovanKraay  for every feature development on spring-data-cosmos - we will need to create 2 PRs, one for main branch, and one for sb3 branch. 
Once this is merged, we can create a new PR for sb3 branch - changes will look exactly the same.

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (kushagraThapar) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java]: @TheovanKraay  - I agree with @FabianMeiswinkel  here. Almost every single production application will have WARN level logs enabled by default. Also, this is a performance regression for sure, as customers are very very sensitive to what gets logged or not. Its better to only log it once per application if possible (you can probably keep a boolean flag to make sure this only gets logged once).

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (trande4884) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:356]: @FabianMeiswinkel  We do not, sub-partitioning is planned for implementation this semester.

**Review by trande4884 — APPROVED** [review]

File: sdk/spring/azure-spring-data-cosmos/CHANGELOG.md
Comment (FabianMeiswinkel) [inline sdk/spring/azure-spring-data-cosmos/CHANGELOG.md:8]: Why is this still technically a breaking change? I don't think it is anymore?

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (FabianMeiswinkel) [inline sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:360]: Please change the error message here to inlcude at least 1 sentence for the motivation like.

The partitionKey is not id!! Consider using findById(ID id, PartitionKey partitionKey) instead to avoid the need for using a cross partition query which results in higher latency and cost than necessary. See https://aka.ms/PointReadsInSpring for more info.

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (FabianMeiswinkel) [inline sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:359]: NIT: This is not thread-safe - you might end-up with multiple logs. Not too critical (maybe even intentional because while possibly logging multiple times the check whether logging happened becomes cheaper) - so I leave it to you whether you decide to make it thread-safe or not.

If you do - change the pointReadWarningLogged to an AtomicBoolean

Then use 
if (this.pointReadWarningLogged.compareAndSet(false, true) {
    ...
}

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM except for few minor comments about log message and changelog text

File: sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java
Comment (TheovanKraay) [reply sdk/spring/azure-spring-data-cosmos/src/main/java/com/azure/spring/data/cosmos/core/CosmosTemplate.java:359]: Discussed already with Kushagra, view was that it’s ok (without even thinking about the cost of thread-safe check). So I think it’s ok as is. Thanks 😊

**Review by kushagraThapar — APPROVED** [review]

**Review by kushagraThapar — APPROVED** [review]
LGMT, great work @TheovanKraay , thank you!!

**Pull request merged**
